core/node/node_framework/src/implementations/resources/pools.rs

```diff
@@ -1,175 +1,148 @@
             tracing::info!(
-                "Created a new master pool. Master pool total connections count: {total_connections}"
+                "Provided a new copy of an existing {} unbound pool",
+                P::kind_str()
             );
+            return Ok(pool.clone());
         }
-
-        result
+        let pool = self.builder().build().await?;
+        *unbound_pool = Some(pool.clone());
+
+        let old_count = self
+            .connections_count
+            .fetch_add(self.max_connections, Ordering::Relaxed);
+        let total_connections = old_count + self.max_connections;
+        tracing::info!(
+            "Created a new {} pool. Total connections count: {total_connections}",
+            P::kind_str()
+        );
+
+        Ok(pool)
     }
 
-    pub async fn get_singleton(&self) -> anyhow::Result<ConnectionPool<Core>> {
+    pub async fn get_singleton(&self) -> anyhow::Result<ConnectionPool<P::DbMarker>> {
         self.get_custom(1).await
     }
 
-    pub async fn get_custom(&self, size: u32) -> anyhow::Result<ConnectionPool<Core>> {
-        let result = self.builder.clone().set_max_size(size).build().await;
+    pub async fn get_custom(&self, size: u32) -> anyhow::Result<ConnectionPool<P::DbMarker>> {
+        let result = self.builder().set_max_size(size).build().await;
 
         if result.is_ok() {
             let old_count = self.connections_count.fetch_add(size, Ordering::Relaxed);
             let total_connections = old_count + size;
             tracing::info!(
-                "Created a new master pool. Master pool total connections count: {total_connections}"
+                "Created a new {} pool. Total connections count: {total_connections}",
+                P::kind_str()
             );
         }
 
         result
     }
 }
 
-/// Represents a connection pool to the replica database.
 #[derive(Debug, Clone)]
-pub struct ReplicaPoolResource {
-    connections_count: Arc<AtomicU32>,
-    builder: ConnectionPoolBuilder<Core>,
-}
-
-impl Resource for ReplicaPoolResource {
-    fn name() -> String {
-        "common/replica_pool".into()
-    }
-}
-
-impl ReplicaPoolResource {
-    pub fn new(builder: ConnectionPoolBuilder<Core>) -> Self {
-        Self {
-            connections_count: Arc::new(AtomicU32::new(0)),
-            builder,
-        }
-    }
-
-    pub async fn get(&self) -> anyhow::Result<ConnectionPool<Core>> {
-        let result = self.builder.build().await;
+#[non_exhaustive]
+pub struct MasterPool {}
 
-        if result.is_ok() {
-            self.connections_count
-                .fetch_add(self.builder.max_size(), Ordering::Relaxed);
-            let total_connections = self.connections_count.load(Ordering::Relaxed);
-            tracing::info!(
-                "Created a new replica pool. Replica pool total connections count: {total_connections}"
-            );
-        }
-
-        result
-    }
+#[derive(Debug, Clone)]
+#[non_exhaustive]
+pub struct ReplicaPool {}
```

Comment:
Bikeshedding: IMO, a struct as a newtype wrapping `()` is a more conventional way to declare marker types. Obviously, doesn't matter much.

Reply:
Makes sense, will do in the next PR.